source/common/filesystem/win32/filesystem_impl.cc

```diff
@@ -102,8 +103,124 @@ std::string InstanceImplWin32::fileReadToEnd(const std::string& path) {
   return file_string.str();
 }
 
+void InstanceImplWin32::splitFileName(std::string& path, std::string& name) {
+  size_t last_slash = path.find_last_of(":/\\");
+  if (last_slash == std::string::npos) {
+    throw EnvoyException(fmt::format("invalid file path {}", path));
+  }
+
+  name.clear();
+  name.append(path.substr(last_slash + 1, std::string::npos));
+
+  // Retain entire single '/', 'd:' drive, and 'd:\' drive root paths
+  // truncate all other trailing slashes
+  path.resize(last_slash +
+              (last_slash == 0 || path[last_slash] == ':' || path[last_slash - 1] == ':'));
+}
+
+// clang-format off
+//
+// Filename warnings and caveats are documented at;
+// https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+//
+// Note special delimiter cases for path prefixes;
+//     "D:\" for local drive volumes
+//     "\server\share\" for network volumes
+//     "\\?\" to pass path directly to the underlying driver
+//          (invalidates the '/' separator and bypasses ".", ".." handling)
+//     "\\?\D:\" for local drive volumes
+//     "\\?\UNC\server\share\" for network volumes (literal "UNC")
+//     "\\.\" for device namespace (e.g. volume names, character devices)
+// File path components must not end in whitespace or '.' (except literal "." and "..")
+// Allow ':' for drive letter only (attempt to name alternate file stream)
+// Allow '/', '\\' as path delimiters only
+// Valid file name character excluding delimiters;
+
+static const char filename_char_table[] = {
```

Comment:
Did you copy this from somewhere? If so can you reference it? In general, is there no win32 function that does all of this? I find that surprising. It would be good if we could use a platform function for this?

Reply:
It would be lovely for Windows to provide this functionality, 30 years later I'm still not holding my breath :)

The logic is almost entirely my own (with minor corrections and clarifications) all licensed under Apache License 2.0;
http://svn.apache.org/viewvc/apr/apr/trunk/file_io/win32/filepath.c?view=annotate
http://svn.apache.org/viewvc/apr/apr/trunk/file_io/win32/filesys.c?view=annotate

I'm not so worried about that citation (being my own code), as opposed citing the article its based on at the Microsoft website, which we already did in this PR.

Reply:
To solve code provenance, I've added the link to my introduction of that table to the APR project.